### PR TITLE
added validation on cidr mask validation

### PR DESCRIPTION
**Pull request by sawyerawesome**
- Added validation on cases where cidr is not respecting the correct format and masks are not contained.
- Added check on network being both included and excluded
 - Added support for IPv6

Fixes: aporeto-inc/aporeto-devs#113

### CODE REVIEW

**Comment by t00f** [comment]
@sawyerawesome we'll need this PR in 3.14 as well

File: netutils/cidr.go [hunk 1]
```diff
@@ -6,11 +6,22 @@ import (
 	"strings"
 )
 
-// prefixIsContained is used to check if an IP is included in a list of CIDRs
-func prefixIsContained(cidrs []*net.IPNet, ip net.IP) bool {
+// ipMaskContains is used to check if a mask's range contains another mask's range
+func ipMaskContains(mask1 net.IPMask, mask2 net.IPMask) bool {
+	for i := 0; i < 4; i++ {
```
Comment (satyamsi) [inline netutils/cidr.go:11]: @sawyerawesome this is not correct. It is not safe to do this. This (hardcoding 4) will not work for IPv6.

Comment (satyamsi) [reply netutils/cidr.go:11]: Use the Size() function on masks to make it more portable.
`
ones1, size1 := mask1.Size()
ones2, size2 := mask2.Size()
if size1 != size 2 {
    return false
}
return ones1 >= ones2
`

File: netutils/cidr_test.go [hunk 2]
```diff
@@ -39,18 +39,26 @@ func Test_prefixIsContained(t *testing.T) {
 			},
```
Comment (satyamsi) [inline netutils/cidr_test.go:39]: Please add IPv6 test cases as well.

File: netutils/cidr.go (hunk 1 above)
Comment (sawyerawesome) [reply netutils/cidr.go]: Yup, should have marked this as not ready for review. I just added support for IPv6 and this issue is addressed

**Comment by sawyerawesome** [comment]
@satyamsi added support for IPv6 and addressed your comments

File: netutils/cidr_test.go [hunk 3]
```diff
@@ -39,18 +39,26 @@ func Test_prefixIsContained(t *testing.T) {
...
 			want: true,
 		},
+		{
+			name: "cidr not contained test failure",
+			args: args{
+				prefixes: []string{"10.10.0.0/16"},
+				ip:       "10.10.0.0/8",
+			},
+			want: false,
+		},
```
Comment (satyamsi) [inline netutils/cidr_test.go:49]:
```

		{
			name: "basic ipv6 test",
			args: args{
				prefixes: []string{"::/0"},
				ip:       "::/0",
			},
			want: true,
		},
		{
			name: "basic ipv6 test 2",
			args: args{
				prefixes: []string{"::/0"},
				ip:       "a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:a0a0/128",
			},
			want: true,
		},
		{
			name: "basic ipv6 test failure",
			args: args{
				prefixes: []string{},
				ip:       "a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:a0a0/128",
			},
			want: false,
		},
		{
			name: "ipv6 multiple match test",
			args: args{
				prefixes: []string{"a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:0000/112", "a0a0:a0a0:a0a0:a0a0:0000:0000:0000:0000/64", "::/0"},
				ip:       "a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:a0a0:a0a0/128",
			},
			want: true,
		},
		{
			name: "ipv6 cidr not contained test failure",
			args: args{
				prefixes: []string{"a0a0:a0a0:a0a0:a0a0::/64"},
				ip:       "a0a0:a0a0::/32",
			},
			want: false,
		},
```

File: netutils/cidr_test.go [hunk 4]
```diff
@@ -87,6 +95,13 @@ func Test_ValidateCIDRs(t *testing.T) {
 			},
 			wantErr: true,
 		},
+		{
+			name: "basic cidr not contained failure test",
+			args: args{
+				[]string{"10.10.10.0/16", "!10.10.10.10/8"},
+			},
+			wantErr: true,
+		},
```
Comment (satyamsi) [inline netutils/cidr_test.go:167]:
```
		{
			name: "ipv6 basic test",
			args: args{
				[]string{"a0a0:a0a0:a0a0:a0a0::/64"},
			},
			wantErr: false,
		},
		{
			name: "ipv6 basic inclusion test",
			args: args{
				[]string{"a0a0:a0a0:a0a0::/48", "!a0a0:a0a0:a0a0:a0a0::/64"},
			},
			wantErr: false,
		},
		{
			name: "ipv6 basic failure test",
			args: args{
				[]string{"!a0a0:a0a0:a0a0:a0a0::/64"},
			},
			wantErr: true,
		},
		{
			name: "ipv6 recursive test",
			args: args{
				[]string{"a0a0:a0a0:a0a0:a0a0::/64", "!a0a0:a0a0:a0a0::/48", "a0a0:a0a0::/32", "!a0a0::/16", "::/0"},
			},
			wantErr: false,
		},
		{
			name: "ipv6 recursive fail test",
			args: args{
				[]string{"a0a0:a0a0:a0a0:a0a0::/64", "!a0a0:a0a0:a0a0::/48", "a0a0:a0a0::/32", "!a0a0::/16"},
			},
			wantErr: true,
		},
		{
			name: "ipv6 recursive multiple not test",
			args: args{
				[]string{"!a0a0:a0a0:a0a0:a0a0::/64", "!a0a0:a0a0:a0a0::/48", "!a0a0:a0a0::/32", "!a0a0::/16", "::/0"},
			},
			wantErr: false,
		},```

**Comment by sawyerawesome** [comment]
@satyamsi @t00f added more test cases

**Comment by sawyerawesome** [comment]
@satyamsi simplified code to use one list saving both ipv4 and ipv6 addresses as you suggested.

**Comment by sawyerawesome** [comment]
@satyamsi @t00f added check for network being both included and excluded.

**Comment by t00f** [comment]
Waiting for @satyamsi approval first before backporting this in 3.14

**Comment by sawyerawesome** [comment]
@satyamsi @t00f added a private cidr struct for better handling and duplication check is added. Also added test cases for duplication checks.

File: netutils/cidr.go [hunk 5]
```diff
@@ -6,61 +6,91 @@ import (
...
 
-	prefixes := []*net.IPNet{}
-	for _, s := range cidrs {
-		_, network, err := net.ParseCIDR(s)
-		if err != nil {
-			return nil, fmt.Errorf("%s is not a valid CIDR", s)
+	var count int
+	for _, pc := range cidrs {
+		if pc.ipNet.String() == net.String() {
+			count++
```
Comment (satyamsi) [inline netutils/cidr.go:52]: if count > 1 return right here. why run the whole loop ?

**Pull request merged**